### PR TITLE
Added sonoff smart water valve equipment

### PR DESCRIPTION
**Pull request by lukezxm**
Since fromzigbee.ts does not have 'msFlowMeasurement', it needs to be added.

### CODE REVIEW

File: src/devices/sonoff.ts [hunk 1]
```diff
@@ -132,6 +133,182 @@ const sonoffExtend = {
...
 
+        return {
+            exposes: [exposes],
+            fromZigbee,
+            toZigbee,
+            isModernExtend: true,
+        };
+    },
+    cyclicTimedIrrigation: (): ModernExtend => {
+        const exposes = e.composite('timedIrrigation', 'cyclic_timed_irrigation', ea.ALL)
```
Comment (Koenkk) [inline src/devices/sonoff.ts:144]:
```suggestion
        const exposes = e.composite('cyclic_timed_irrigation', 'cyclic_timed_irrigation', ea.ALL)
```

File: src/devices/sonoff.ts [hunk 2]
```diff
@@ -132,6 +133,182 @@ const sonoffExtend = {
...
+            .withDescription('Smart water valve cycle timing irrigation')
+            .withFeature(e.numeric('currentCount', ea.STATE).withDescription('Number of times it has been executed').withUnit('times'))
```
Comment (Koenkk) [inline src/devices/sonoff.ts:146]:
```suggestion
            .withFeature(e.numeric('current_count', ea.STATE).withDescription('Number of times it has been executed').withUnit('times'))
```

Make all the features also `snake_case`

File: src/devices/sonoff.ts [hunk 3]
```diff
@@ -132,6 +133,182 @@ const sonoffExtend = {
...
+                    meta.logger.debug(`totalNumberOfTimesBuffer ${totalNumberBuffer}`);
+                    meta.logger.debug(`irrigationDurationBuffer ${irrigationDurationBuffer}`);
+                    meta.logger.debug(`irrigationIntervalBuffer ${irrigationIntervalBuffer}`);
+
+                    return {
+                        cyclic_timed_irrigation: {
+                            currentCount: currentCountBuffer,
+                            totalNumber: totalNumberBuffer,
+                            irrigationDuration: irrigationDurationBuffer,
+                            irrigationInterval: irrigationIntervalBuffer,
```
Comment (Koenkk) [inline src/devices/sonoff.ts:181]:
```suggestion
                            irrigation_interval: irrigationIntervalBuffer,
```

Also update these

File: src/devices/sonoff.ts [hunk 4]
```diff
@@ -649,6 +826,46 @@ const definitions: Definition[] = [
...
+                skipDuplicateTransaction: true,
+                configureReporting: true,
+            }),
+            sonoffExtend.cyclicTimedIrrigation(),
+            sonoffExtend.cyclicQuantitativeIrrigation(),
+        ],
+        configure: async (device, coordinatorEndpoint, logger) => {
+            const endpoint = device.getEndpoint(1);
+            await reporting.bind(endpoint, coordinatorEndpoint, ['genPowerCfg']);
+            await reporting.bind(endpoint, coordinatorEndpoint, ['msFlowMeasurement']);
```
Comment (Koenkk) [inline src/devices/sonoff.ts:864]: Please create a new modern extend for this here https://github.com/Koenkk/zigbee-herdsman-converters/blob/bc0506cd8e40d7c4ec5b108cc2c652c7b184dead/src/lib/modernExtend.ts#L743

File: src/devices/sonoff.ts [hunk 5]
```diff
@@ -649,6 +826,46 @@ const definitions: Definition[] = [
...
+        toZigbee: [],
+        ota: ota.zigbeeOTA,
+        exposes: [
+            e.battery(),
+            e.numeric('flow', ea.STATE).withDescription('Current water flow').withUnit('m³/h'),
+        ],
+        extend: [
+            enumLookup({
+                name: 'current_device_status',
+                lookup: {'normal state': 0, 'water shortage': 1, 'water leakage': 2, 'water shortage & water leakage': 3},
```
Comment (Koenkk) [inline src/devices/sonoff.ts:847]: Please replace spaces with `_` e.g. `water_shortage`

**Comment by lukezxm** [comment]
All changes have been made as suggested, thanks

File: src/devices/sonoff.ts [hunk 6]
```diff
@@ -649,6 +826,46 @@ const definitions: Definition[] = [
...
+                configureReporting: true,
+            }),
+            sonoffExtend.cyclicTimedIrrigation(),
+            sonoffExtend.cyclicQuantitativeIrrigation(),
+        ],
+        configure: async (device, coordinatorEndpoint, logger) => {
+            const endpoint = device.getEndpoint(1);
+            await reporting.bind(endpoint, coordinatorEndpoint, ['genPowerCfg']);
+            await reporting.bind(endpoint, coordinatorEndpoint, ['msFlowMeasurement']);
+            await endpoint.read('genPowerCfg', ['batteryPercentageRemaining']);
```
[the remark this answers is not in the document]
Comment (Koenkk) [reply src/devices/sonoff.ts]: @lukezxm this change is still missing

**Comment by lukezxm** [comment]
Hello! When will my push be merged into the official version?

**Comment by Koenkk** [comment]
Missed it, thanks for pinging me!

**Pull request merged**